### PR TITLE
feat: add strategy registry

### PR DESCRIPTION
**Pull request by pandadefi**

### CODE REVIEW

File: contracts/StrategyVersionRegistry.vy [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+@external
+def setGovernance(governance: address):
+    """
+    @notice Starts the 1st phase of the governance transfer.
+    @dev Throws if the caller is not current governance.
+    @param governance The next governance address
+    """
+    assert msg.sender == self.governance  # dev: unauthorized
+    log NewPendingGovernance(msg.sender)
```
Comment (poolpitako) [inline contracts/StrategyVersionRegistry.vy:44]: `governance` instead of `msg.sender`?

File: contracts/StrategyVersionRegistry.vy [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+def addNewRelease(strategy :address, name: String[64] = ""):
+    """
+    @notice add a new version of a strategy
+    """
+    assert msg.sender == self.governance  # dev: unauthorized
+    assert Strategy(strategy).isOriginal() # dev: not original 
+    strategyName: String[64] = name 
+    if name == "":
+        strategyName = Strategy(strategy).name()
+        assert strategyName != ""
```
Comment (poolpitako) [inline contracts/StrategyVersionRegistry.vy:70]: can't you reuse `name`? why do you need `strategyName`?

File: contracts/StrategyVersionRegistry.vy [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+@view 
+@external
+def latestRelease(name: String[64], apiVersion: String[8]) -> address:
+    key: String[73] = self._computeKey(name, apiVersion)
+    return self.strategyVersions[key]
+
+@external
+def clone(strategy: address, params: Bytes[256]):
+    assert Strategy(strategy).isOriginal()  # dev: not original
```
Comment (poolpitako) [inline contracts/StrategyVersionRegistry.vy:84]: you are checking that the strategy is original but you are not checking if the strategy belongs to the registry. why?

File: contracts/StrategyVersionRegistry.vy (hunk 2 above)
Comment (pandadefi) [reply contracts/StrategyVersionRegistry.vy:70]: That's not possible, I get a `ImmutableViolation: Cannot write to calldata` if I try to reuse name

File: contracts/StrategyVersionRegistry.vy (hunk 3 above)
Comment (pandadefi) [reply contracts/StrategyVersionRegistry.vy:84]: The strategy doesn't have to be in the registry to be cloned. Registry helps out finding which strategy is the latest and then you can call the clone function with the address, but you can also clone a strategy that's not part of the registry..

WDYT?

File: contracts/StrategyVersionRegistry.vy (hunk 1 above)
Comment (pandadefi) [reply contracts/StrategyVersionRegistry.vy:44]: yes

File: contracts/StrategyVersionRegistry.vy (hunk 2 above)
Comment (poolpitako) [reply contracts/StrategyVersionRegistry.vy]: interesting!

File: contracts/StrategyVersionRegistry.vy (hunk 3 above)
Comment (poolpitako) [reply contracts/StrategyVersionRegistry.vy:84]: I believe only strategies in the registry should be clonable. The same way the vault registry works.

Comment (pandadefi) [reply contracts/StrategyVersionRegistry.vy:90]: ok, I have made some checks but since you can put a custom name the check can be bypassed.

**Comment by pandadefi** [comment]
closing this for now

**Pull request closed without merging**